query/src/datasources/fuse_table/io/block_appender.rs

```diff
@@ -51,113 +49,113 @@ where T: MetaApi + Send + Sync + 'static
-                    })
-                })
-                .collect::<HashMap<_, _>>();
+                .map(|(idx, v)| (*idx, v.1.clone()))
+                .collect::<HashMap<ColumnId, ColStats>>();
 
             let block_info = BlockMeta {
                 location: BlockLocation {
                     location: location.clone(),
                     meta_size,
                 },
                 row_count,
-                block_size,
+                block_size: block_in_memory_size,
                 col_stats,
             };
 
-            blocks.push(block_info);
+            block_metas.push(block_info);
+            blocks_stats.push(blk_stats);
 
             summary_block_count += 1;
             summary_row_count += row_count;
-            summary_compressed_byte_size += block_size;
-            summary_uncompressed_byte_size += rg_meta.total_byte_size() as u64;
-            block_stats.push(col_dyn_stats);
-
-            // write to storage
-            self.data_accessor(&ctx)?
-                .put(&location, cursor.into_inner())
-                .await?;
-        }
-
-        let summary = reduce(block_stats);
-
-        let mut res = HashMap::new();
-        for (col, col_stats) in summary {
-            match col_stats {
-                Ok(Some(stats)) => {
-                    let s = serialize_statistics(stats.as_ref());
-                    let c = ColStats {
-                        min: s.min_value,
-                        max: s.max_value,
-                        null_count: s.null_count.map(|v| v as u64),
-                    };
-                    res.insert(col, c);
-                }
-                _ => todo!(),
-            }
+            summary_compressed_byte_size += file_size;
+            summary_uncompressed_byte_size += block_in_memory_size;
         }
 
+        let summary = column_stats_reduce(blocks_stats)?;
         let segment_info = SegmentInfo {
-            blocks,
+            blocks: block_metas,
             summary: Stats {
                 row_count: summary_row_count,
                 block_count: summary_block_count,
                 uncompressed_byte_size: summary_uncompressed_byte_size,
                 compressed_byte_size: summary_compressed_byte_size,
-                col_stats: res,
+                col_stats: summary,
             },
         };
         Ok(segment_info)
     }
 }
 
-pub(crate) fn write_in_memory(arrow_schema: &ArrowSchema, block: DataBlock) -> Result<Vec<u8>> {
+pub fn block_stats(data_block: &DataBlock) -> Result<HashMap<ColumnId, (DataType, ColStats)>> {
+    // TODO column id is FAKED, this is OK as long as table schema is NOT changed, which is not realistic
+    // we should extend DataField with column_id ...
+    (0..).into_iter().zip(data_block.columns().iter()).try_fold(
+        HashMap::new(),
+        |mut res, (idx, col)| {
+            let data_type = col.data_type();
+            let min = match col {
```

Comment:
So any column will have `min`, `max` stats?

Reply:
Right.
This will be used for our background merging and automatic clustering.
I think @dantengsky will introduce it later.